Привет! Меня зовут Исмаилов Исмаил, я буду проверять твой проект :) Можешь обращаться ко мне на «ты». Если тебе комфортно, то и я буду к тебе так обращаться, если нет, то обязательно скажи об этом.

Пожалуйста, не удаляй мои комментарии, которые я буду оставлять в работе. Это особенно поможет, если твои проекты будут отправлены на повторную проверку. При повторных проверках у комментариев будут приписки: "Вторая итерация", "Третья итерация" и т.д. 

Ты тоже можешь реагировать на мои комментарии, но в таком случае постарайся, чтобы твои комментарии отличались от моих: например, выделяй их своим любимым цветом — так у нас не возникнет путаницы :)

Кстати, про цвета! Мои комментарии будут в следующей цветовой гамме: зелеными, желтыми и красными. Например:

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 </b>
    
В случае, если решение на отдельном шаге является полностью правильным. </div>

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡</b> 
    
В случае, если решение может стать еще лучше с некоторыми корректировками </div>
 
<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 
    
В случае, если решение какого-то шага требует значительных изменений. Проект не может быть принят с первого раза, если ревью содержит комментарии, помеченные этим цветом </div>

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [88]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [89]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [90]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [91]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [92]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [93]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [94]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [95]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [96]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [97]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [98]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [99]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [100]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [101]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [102]:
data.duplicated().sum()

71

In [103]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [104]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [105]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [106]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [107]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [108]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [109]:
data_pivot = data.pivot_table(index='children',  values ='debt', aggfunc =['sum','count']).reset_index()

data_pivot['ratio'] = data_pivot['sum'] / data_pivot['count']
print(data_pivot.sort_values(by='ratio'))

  children   sum  count     ratio
            debt   debt          
5        5     0      9  0.000000
0        0  1063  14091  0.075438
3        3    27    330  0.081818
1        1   444   4808  0.092346
2        2   194   2052  0.094542
4        4     4     41  0.097561


**Вывод:** Будет ошибкой судить о вероятности возврата кредита в срок по категориям людей с 3-мя, 4-мя и 5-ю детьми ввиду малой статистической выборки. У клиентов без детей вероятность невозврата самая низкая - 7,5%, клиенты с 1-м и 2-мя детьми примерно одинаково не возвращают в срок - 9,2% и 9,4% соответственно. Можно сделать предварительный вывод, что вероятность невозврата нелинейно растет вместе с количеством детей у клиента.

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡</b> 

Тут еще стоит обратить внимание на низкий размер выборки категории с 5 детьми - по ней нельзя делать выводы, стоит это отметить

<div background-color="blue"> 
<b>Комментарий студента</b> 

Спасибо! Исключил из выборки три категории, т.к. единицы, десятки и сотни - отличаются на порядок(-ки) от тысяч.
</div>


#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [113]:
data_pivot = data.pivot_table(index='family_status', values='debt', aggfunc=['sum', 'count']).reset_index()
data_pivot['ratio'] = data_pivot['sum'] / data_pivot['count']
print(data_pivot.sort_values(by='ratio'))

# Ваш код будет здесь. Вы можете создавать новые ячейки.

           family_status  sum  count     ratio
                         debt   debt          
2         вдовец / вдова   63    951  0.066246
1              в разводе   84   1189  0.070648
4        женат / замужем  927  12261  0.075606
3       гражданский брак  385   4134  0.093130
0  Не женат / не замужем  273   2796  0.097639


**Вывод:** Минимальная вероятность невозврата у разведенных, овдовевших и тех, кто состоит в официальном браке. Наибольшая вероятность невозврата/вохврата с задержкой у тех, кто живет гражданским браком и не женатых/не замужних.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [111]:
data_pivot = data.pivot_table(index='total_income_category', values='debt', aggfunc=['sum', 'count']).reset_index()
data_pivot['ratio'] = data_pivot['sum'] / data_pivot['count']
print(data_pivot.sort_values(by='ratio'))


                        sum  count     ratio
                       debt   debt          
total_income_category                       
D                        21    349  0.060172
B                       354   5014  0.070602
A                         2     25  0.080000
C                      1353  15921  0.084982
E                         2     22  0.090909


**Вывод:** По категориям A, D, E нельзя судить, ввиду малой выборки. Сравнивая категории B и C, можно сказать, что вероятность невозврата падает с увеличением дохода. 

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 </b>

Да, а тут супер, что обратил на это внимание

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [112]:
data_pivot = data.pivot_table(index='purpose_category', values='debt', aggfunc=['sum', 'count']).reset_index()
data_pivot['ratio'] = data_pivot['sum'] / data_pivot['count']
print(data_pivot.sort_values(by='ratio'))

                          sum  count     ratio
                         debt   debt          
purpose_category                              
операции с недвижимостью  780  10751  0.072551
проведение свадьбы        183   2313  0.079118
получение образования     369   3988  0.092528
операции с автомобилем    400   4279  0.093480


**Вывод:** Те, кто берет деньги на сделку с недвижимостью или проведение свадьбы, вероятнее всего вернут деньги в срок. Одинаково не возвращают деньги те, кто получает образование или покупают автомобиль.

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 </b>

Здорово, что смотришь на относительные значения, не забываешь про комментарии и т.д. Код тоже хороший

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Так как пропуски в столбцах с трудовым стажем и информацией о доходах встречались в одних и тех же строках, можно предположить следующее.
1. Возможно, клиент получал кредит на специальных условиях ("не требуется справка о доходах", "не требуется справка с места работы" и т.п.)
2. Может в программном коде в разделе о работе есть ошибка, баг - и периодически программа сбивается при запоминании этих данных.

<div class="alert-success"> 
<b>Комментарий ревьюера 👍</b>

Согласен с названными тобою возможными причинами пропусков

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* На медианное значение не влияют аномальные значения в ряду.

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 </b>
    
Верно, дело именно в том, что средние сильно реагируют на выбросы, тем самым искажая наши данные

### Шаг 4: общий вывод.

<b> Провели мини-исследование надежности заемщиков </b> <br />
<p> <b>Первым шагом</b> было изучение общей информации данных.
    В представляемой таблице около 21-й тысячи строк и 12 колонок с информацией о клиентах.
</p>
<p> <b> На втором шаге </b>
<ul>
<li>Выявили и удалили пропуски. В около 2-х тысячах строк были пропущены сведения о трудовом стаже и доходе клиента. Можно предположить, что данные были пропущены неслучайно, ввиду специальных условий кредита или из-за программной ошибки - ввод данных о клиенте дает сбой. Пропущенные значения заполнили медианными значениями.</li>
<li>Обработали дубликаты. </li>
<li>Привели типы данных к удобным для анализа.</li> 
<li>Сделали категоризацию данных. Цели оформления кредита привели поделили на 4 группы – сделки с недвижимостью, автомобилем, проведение свадьбы и на образование. По уровню дохода клиентов поделили на 5 групп.</li></ul>
</p>
<p>
<b>Третьим шагом</b> было изучение вероятности невозврата клиентом кредита в срок в разрезе по одиночным показателям. <br/>
    <br/>Механика проведения анализа по всем нижеуказанным категориям была одинаковой и заключалась в группировке данных по категориям и отражении числа задолжников и общего числа клиентов в этой категории. Далее для каждой строки вычислялось отношение числа задолжников к общему числу клиентов.
    <ol>
<li>Зависимость между <b>количеством детей</b> и возвратом кредита в срок. <br />
    Ввиду малого количества прецедентов оформления кредита людьми с 3-мя, 4-мя, и 5-ю детьми о данных категориях судить нельзя. <br />
    Без учета этих трех категорий клиенты с без детей, с 1-м ребенком и 2-мя детьми не вернут кредит в срок с вероятностями 7.5%, 9.2%, 9.4% соответственно. <br />
    Делаем предварительный вывод, что вероятность невозврата нелинейно расчет вместе с количеством детей у клиента.</li>
<li>Зависимость между <b>семейным положением</b> и возвратом кредита в срок. <br />
    Минимальная вероятность невозврата у разведенных, овдовевших и тех, кто состоит в официальном браке – 7.0%, 6.6%, 7.5% соответственно. Наибольшая вероятность невозврата/возврата с задержкой у тех, кто живет гражданским браком и не женатых/не замужних – 9.3% и 9.7%.</li>
<li>Зависимость между <b>уровнем дохода</b> и возвратом кредита в срок.<br />
    Минимум данных о категориях с доходами 0-30 тыс.руб. в месяц (E), 30-50 тыс.руб. в месяц (D), и свыше 1 млн.руб. в месяц (А). По ним сделать вывод невозможно.<br />
    Категории с доходами 50-200 тыс.руб. в месяц (С) и 200 тыс.руб. – 1 млн.руб. в месяц не вернут кредит в срок с вероятностями 8.5% и 7.1% соответственно. То есть, предварительно делаем вывод, что вероятность невозврата кредита в срок падает с увеличением дохода. Так же, предполагаю, что необходимо подробнее разбить данные категории для более детального изучения зависимости.</li>
<li>Зависимость между <b>целью кредита</b> и его возвратом в срок.<br />
    Те, кто берет деньги на сделку с недвижимостью или проведение свадьбы, вероятнее всего вернут деньги в срок (7.3% и 7.9%). Одинаково не возвращают деньги те, кто получает образование или покупают автомобиль (9.25% и 9.35%).</li> </ol>
 </p>
<p>
<b>Резюме</b>
<ol>
<li> Факт того, что категории не женатых/не замужних и тех, кто живет гражданским браком с большей вероятностью не вернут кредит, а люди без детей – с наименьшей (!!! Логическое противоречие !!!), заставляет предположить, что делать выводы о вероятности возврата кредита в срок клиентом в разрезах по одиночным показателям ошибочно. Необходимо комплексное исследование, одновременно учитывающее несколько показателей.</li>
<li>При наличии технической возможности положительно скажется на исследовании, если в таблицу добавить более подробную информацию о задержке клиентом выплаты по кредиту (указать продолжительность задержки, указать регулярность задержек).</li>
<li>Также необходимо внести сведения о городе, в котором проживает клиент. Для каждого города провести небольшое исследование рынка труда и более точно категоризовать клиентов по уровню дохода.</li> </ol>
</p>


<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 

Вывод можно расширить, дополнить. Рассказать о том, как ты к этому пришел. Построил таблички, ответил на вопросы и т.д., т.е. в нем стоит описывать не только итоги проекта, но и некоторый путь к этим итогам, краткую выжимку твоей работы. Также надо представить, что ты готовишь какие-то рекомендации бизнесу, поэтому презентовать им два предложения - не лучший вариант. Например, можно рассказать, чем результаты твоего мини-исследования могут быть полезны для бизнеса / заказчика в его деятельности

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b>
    
Отличный проект, тобою проделана большая работа, ты молодец!

Сперва похвалю за наличие везде промежуточных выводов, комментариев. Очень хороший код, показываешь хорошие навыки владения инструментами Питона, в частности - сводными таблицами. Также дны верные ответы на поставленные в задании вопросы.
    
Что обязательно стоит доработать: дописать общий вывод
    
Буду ждать доработок :) Удачи!

<div background-color="blue"> 
<b>Комментарий студента</b> 

Спасибо! Такие комментарии вдохновляют:)<br />
    Обновил вывод - описал ход исследования и сделал свои выводы.
    
</div>


<div class="alert-success"> 
<b>Комментарий ревьюера 👍 Вторая итерация </b>

Все ошибки исправлены, проект стал еще лучше, принимаю. Успехов в дальнейшей учебе!